src/Oscoin/Storage/Block/Orphanage.hs

```diff
@@ -270,17 +297,99 @@ storeOrphan
+    Candidate    |                         |                            |
+            +-+  |                         |                            |
+            | | /\                         \                            \
+            | |-  -\                        -\                           -\
+            |<|     >                         >                            >
+            |||
+            |||
+            |v|
+            |-|                       /-
+          /-|||                     /-
+        /-  |||                   /-
+      <-    |||                 <-
+            |v|                         \
+            | |\                         \
+            | | v                         v
+            | |
+            +-+
+
+------------------------------------------------------------------------------}
+
+-- | Like 'pruneOrphanageShallow', but also prunes any other candidate chain
+-- branching off from /any/ block of the chain being pruned.
+pruneOrphanageDeep
+    :: (Ord (BlockHash c)
+       , Eq s
+       )
+    => BlockHash c
+    -> ChainCandidate c s
+    -> Orphanage c tx s
+    -> Orphanage c tx s
+pruneOrphanageDeep rootHash bestFork@ChainCandidate{candidateChain} o =
+    foldl' (flip bulkDelete)
+           (pruneOrphanageShallow rootHash bestFork o)
+           (toList candidateChain)
+
 -- | Given a candidate which won, delete the orphans from the store and
--- the chain from the candidates
-pruneOrphanage
-    :: Ord (BlockHash c)
+-- the chain from the candidates.
+pruneOrphanageShallow
+    :: forall c tx s.
+       ( Ord (BlockHash c)
+       , Eq s
+       )
     => BlockHash c
     -> ChainCandidate c s
     -> Orphanage c tx s
     -> Orphanage c tx s
-pruneOrphanage rootHash ChainCandidate{candidateChain} o =
-    let o' = o { orphans = foldl' (flip Map.delete) (orphans o) (toList candidateChain) }
-    in bulkDelete rootHash o'
+pruneOrphanageShallow rootHash inputCandidate@ChainCandidate{candidateChain} o =
+    let (candidates', canPruneOrphans) = pruneCandidate (candidates o)
+        orphans' = if canPruneOrphans
+                      then foldl' (flip Map.delete) (orphans o) (toList candidateChain)
+                      else orphans o
+    in o { orphans = orphans'
+         , candidates = candidates'
+         , tips = tips' (tips o)
+         }
+  where
+      -- Prunes the input 'ChainCandidate'. If this was the only candidate
+      -- for the root hash, also prune the orphan blocks.
+      -- N.B (adn) Is this sufficient or we need full-flow reference counting?
+      pruneCandidate ::  Map (ChainRoot c) (Candidates c s)
+                     -> (Map (ChainRoot c) (Candidates c s), Bool)
+      pruneCandidate old =
+          case map removeCandidate (Map.lookup rootHash old) of
+            Nothing -> (old, False)
+            Just [] -> (Map.delete rootHash old, True)
+            Just xs -> (Map.insert rootHash xs old, False)
```

Comment:
Shouldn’t we return `True` here according to your diagram? If we remove a chain candidate we should also remove all blocks associated with it.

Reply:
The problem is that if you _always_ remove all the orphans (this was the case before) if you had any other chain candidate referencing those orphans, it would be wrong to remove them, and this why I am returning `False` here in case we do still have candidates.

However, while talking with you, I think this is a problem only if you remote the "root hash" (which candidates might share) but it's actually fine if delete only the entries pointed by the candidate. I will have a try and see how it turns out.

As I have written in my comment somewhere, in theory the only way we would be sure we are not deleting an orphan which is shared by multiple chains is by using a reference-counting mechanism for the actual orphans, but that complicates things a little bit so I have refrained from doing it just yet as part of this PR.

Reply:
I see. But doesn’t this result in memory leaks because we don’t garbage collect the blocks that were only referenced by the input candidate?

(This brings me to a related question: Why we use a manual pointer scheme with block hashes. Why doesn’t `candidateChain` hold the blocks and we get rid of `orphans`?)

Reply:
> I see. But doesn’t this result in memory leaks because we don’t garbage collect the blocks that were only referenced by the input candidate?

In the current scheme, that would be caught by the `Just []` case, where we would yield `True` and thus delete all the orphans.

> (This brings me to a related question: Why we use a manual pointer scheme with block hashes. Why doesn’t candidateChain hold the blocks and we get rid of orphans?)

I am not opposed to have the blocks directly inside each `candidateChain`. The reason why I didn't want to do that is that I was concerned you could have many chains referencing the same candidate, and a block can potentially be big (say, max 1MB in case we are very, very unlucky). Eventually I have settled on this design as it's more contrived to work with, but should also be more memory efficient (I hope!)

Reply:
Thanks for clarifying. I think it’s nothing to be addressed right away.